### PR TITLE
Feature advanced scripting

### PR DESCRIPTION
**Pull request by xoores**
Continuing from discussion under #906:

This version can override several fields: urgency, icon_path, stack_tags, category, format, body, summary, progress, transient and timeout. It uses some bash incantations to acheve this, mmap to move data around and some pointer-based parsing.

I tried to document as much as possible to make it easier to understand what I was going for.

If script returns anything else than 0, the notification is thrown away. If one script returns non-zero status, no other scripts are run. This is for debate, someone might want to use this as e.g. way to push notifications to Slack or something and it this scenario it would be of course bad.

If script fails to run or if script outputs too much, it will be ignored (warn will be printed). Meaning no changes will be made to the notification - playing it safe here :slightly_smiling_face: 

I run this version and I use it to change some fields like this:
```bash
#!/bin/bash

if [[ "${DUNST_BODY}" == *mattermost.xxx.yy* ]]; then
       	DUNST_BODY="$(echo "${DUNST_BODY}" | sed -e 's|<a[^@]*||g' -e 's|\(@[^:]*:\)|<b>\1</b>|g' | xargs)"
fi
```

Notes:
- It might be nice to specify some kind of timeout for the script since it is blocking
- script arguments should be abolished as there was no escaping etc (unpredictable)
- Some methods should be moved around - I wanted to edit only the files I had to (easier merging)
- Still have to finish fields such as `colors` as well :shrug:

I also changed the following:
- Changed `notification_print` so it uses LOG_* in more compact way
- Urgency-based colors are applied by default when urgency is changed through script or rule. If something changes color,  urgency-based defaults will not be applied automatically.

Possibly related issues that might be addressed/solved by this: #899, #1012

Questions:
- Should I continue altering `script` or should I create some other option for this so it is 100% backwards compatible?

### CODE REVIEW

**Comment by xoores** [comment]
I fixed the -Wpedantic stuff in e73724ca8c7958d43f5523712c1f2306696f98c4, co next workflow should be fine :slightly_smiling_face:

**Comment by xoores** [comment]
Heh, nothing like a forgotten free :facepalm:

I'll have to figure out why Valgrind doesn't work on my computer, because I already tried to run it exactly because this... I think I found the leak but without Valgrind cannot be sure that it was the only one I've managed to overlook. Will push momentarily.

File: src/notification.c [hunk 1]
```diff
@@ -516,6 +831,31 @@ void notification_init(struct notification *n)
 
 static void notification_format_message(struct notification *n)
 {
+        // Reapply colors based on urgency - this bothered me for quite a long time...
```
Comment (fwsmit) [inline src/notification.c:902]: We can discuss this one in a separate PR/issue as well. I'm not sure if I agree with you here. Two things:
- This approach doesn't work, since the colours are not always stored in the settings struct. They may also be changed by rules and are stored elsewhere
- This is intended behaviour. You are supposed to put your rules that change the urgency above the rules that change the colours based on the urgency.

But feel free to discuss further in a separate PR / issue.

**Comment by xoores** [comment]
Yeah, the issue was not me not knowing how to use Valgrind but rather AVX512 instructions that my CPU supports and Valgrid does not :slightly_smiling_face:

When you try to run it with these instructions not disabled, it will look something like this:
```
==14025== Memcheck, a memory error detector
==14025== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==14025== Using Valgrind-3.20.0.GIT and LibVEX; rerun with -h for copyright info
==14025== Command: ./main
==14025== 
vex amd64->IR: unhandled instruction bytes: 0xC5 0xFD 0x47 0xC0 0xC5 0xE5 0x47 0xDB 0xC5 0xDD
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=1 VEX.L=1 VEX.nVVVV=0x0 ESC=0F
vex amd64->IR:   PFX.66=1 PFX.F2=0 PFX.F3=0
==14025== valgrind: Unrecognised instruction at address 0x401954e.
==14025==    at 0x401954E: _dl_sysdep_start (in /lib64/ld-linux-x86-64.so.2)
==14025==    by 0x401B247: _dl_start (in /lib64/ld-linux-x86-64.so.2)
```
And while I rather quickly identified those as KMOV, I needed to re-compile my toolchain and libraries - it took mi a while :slightly_smiling_face: I got it running this morning.

```
5 tests - 5 passed, 0 failed, 0 skipped (12153 ticks, 0.012 sec)

Total: 183 tests (6465904 ticks, 6.466 sec), 2326 assertions
Pass: 183, fail: 0, skip: 0.
```

That const warning is PITA... I would maybe suggest to add `-Werror` to `test-valgrind` target? It would be much easier not having to juggle these flags around, especially if these are required for test to pass :slightly_smiling_face: 

Fixed along with proposed change to the test-valgrind in 9acbfefc2a79eec3a6259a8ef40e0aa8f3f663b9

**Review by fwsmit — COMMENTED** [review]
I've taken an initial look at the code. Most things made sense, but I'm still trying to figure out if this is the right approach. You are implementing some basic IPC, which makes me wonder if you're not better off using dbus (we are already using it extensively, so why not more?). We could add dunstctl commands like `dunstctl update ID summary "new summary"`.  I'm not saying that this is neccesarily the best approach, but we should at least think about it.

I get that you are seeing a lot of other small things you want to improve, but I think that's best to do in a separate PR to keep this one focused on the script changes.

File: .gitignore [hunk 2]
```diff
@@ -17,3 +17,5 @@ vgcore.*
 /dunst.systemd.service
 /org.knopwob.dunst.service
 /test/test
+
+.idea/
```
[the remark this answers is not in the document]
Comment (xoores) [reply .gitignore:21]: I use CLion from IntelliJ. It uses this directory to store project-specific settings :)

File: src/notification.c [hunk 3]
```diff
@@ -47,130 +48,444 @@ struct _notification_private {
         gint refcount;
 };
 
+extern char **environ;
+
+
+//TODO: Move to utils.c but for now I wanna edit only the files I have to...
```
[the remark this answers is not in the document]
Comment (xoores) [reply src/notification.c:54]: OK, will move it in the next commit. I was pushing most of the changes to one file, because I was not sure if you would want to merge it :slightly_smiling_face:

**Review by xoores — COMMENTED** [review]
Added some responses

**Comment by xoores** [comment]
> ...wonder if you're not better off using dbus (we are already using it extensively, so why not more?). We could add dunstctl commands like `dunstctl update ID summary "new summary"`. I'm not saying that this is neccesarily the best approach, but we should at least think about it.

I thought about it initially but the there are few things that speak against it IMO:
- You still need to somehow hook the notification reception and make dunst wait for response to another DBUS event
- Making DBUS listener+responder is quite hard in simple BASH script, it is not notification based etc. 

We could, of course, create whole framework for DUNST interaction, much as i3 has one but I'd say that it would be completely different beast to do :slightly_smiling_face: It would be much more complex to implement (but OK, one-time job) but also much more demanding on someone who just wants to do some dynamic changes of few things in the notification itself. Instead of (relatively) simple BASH that almost everyone can use and hack something together, you'd end up atleast with Python.

I personally would say that BASH interface is much more user inviting - we could create few sample scripts so BFU can hack these for start. Much easier than Python/C/whatever else :slightly_smiling_face:

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (xoores) [reply .gitignore:21]: I use CLion from IntelliJ. It uses this directory to store project-specific settings :)

File: src/notification.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (xoores) [reply src/notification.c:54]: OK, will move it in the next commit. I was pushing most of the changes to one file, because I was not sure if you would want to merge it :slightly_smiling_face:

File: src/notification.c (hunk 1 above)
Comment (xoores) [reply src/notification.c:902]: Ah, I see - I have the _critical/_normal sections right after global and I append my rules under that. I did not know that this was the intended way to do this.

How would you suggest to do this? Instinctively I expect that if I set urgency to e.g. CRITICAL, it will look like one :slightly_smiling_face:

Perhaps I can re-run rules application after the script change?

File: src/notification.c [hunk 4]
```diff
@@ -353,17 +371,23 @@ int __notification_run_script(struct notification *n, bool blocking, char *calle
...
-                                  PROT_READ|PROT_WRITE,
-                                  MAP_SHARED|MAP_ANONYMOUS,
-                                  -1, 0);
+        script_output = (unsigned char *) mmap(0, SCRIPT_BUFFER_LEN,
+                                               PROT_READ|PROT_WRITE,
+                                               MAP_SHARED|MAP_ANONYMOUS,
+                                               -1, 0);
+
+        if( script_output == MAP_FAILED ) {
+            LOG_E("Failed to allocate %d bytes of shared memory for script output!", SCRIPT_BUFFER_LEN);
```
[the remark this answers is not in the document]
Comment (xoores) [reply src/notification.c]: Yeah, I noticed some unhandled allocation fails throughout the code so I was wandering what was your approach... I'd say that if the system does not have few k of free RAM we are stuffed either way :laughing:

**Comment by xoores** [comment]
I've applied some of your suggestions - the rules make sense now... My fault - I thought that since example rules were at the end, I should just continue and leave `urgency_*` in the middle. Moved them to the end of config and works flawlessly!

Also I removed the crash when alloc fails.

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (liskin) [reply .gitignore:21]: Might be better to put `.idea` into your global `~/.config/git/ignore` then.

[the remark this answers is not in the document]
Comment (xoores) [reply .gitignore:21]: I did not push this commit as it was purely cosmetic and I was sure there will be something else :laughing: And I was right, as I changed the test/test target, it inevitably broke the clang build. At first I did not really pay attention but I went through it as I saw you did something and all builds have passed. My bad.

It will be gone momentarily as I fix the clang build :slightly_smiling_face:

File: src/notification.c (hunk 1 above)
Comment (fwsmit) [reply src/notification.c]: No, the rules should not be reapplied. The script rule should just work like any other rule. This is way easier to implement. I do agree that it should be documented better in the example dunstrc. In fact, I can do that now.

Note as well that reapplying the rules would create an infinite loop of running scripts and reapplying the rules :p

**Comment by fwsmit** [comment]
If you want to use `-Werror` in your own evironment, you can just set `EXTRA_CFLAGS=-Werror` before running make. I would not mess with the Makefile for that.

File: src/notification.c (hunk 1 above)
Comment (xoores) [reply src/notification.c]: I removed it - will push in few minutes :slightly_smiling_face: As I said, the main source was my misunderstanding of the rules.

After this I would probably suggest to maybe make it a little bit cleaner in the doc. It is obvious now as I know how it works, but initialy from the user perspective I really did not get it :slightly_smiling_face:

**Comment by fwsmit** [comment]
> > ...wonder if you're not better off using dbus (we are already using it extensively, so why not more?). We could add dunstctl commands like `dunstctl update ID summary "new summary"`. I'm not saying that this is neccesarily the best approach, but we should at least think about it.
> 
> I thought about it initially but the there are few things that speak against it IMO:
> 
>     * You still need to somehow hook the notification reception and make dunst wait for response to another DBUS event
> 
>     * Making DBUS listener+responder is quite hard in simple BASH script, it is not notification based etc.
> 
> 
> We could, of course, create whole framework for DUNST interaction, much as i3 has one but I'd say that it would be completely different beast to do slightly_smiling_face It would be much more complex to implement (but OK, one-time job) but also much more demanding on someone who just wants to do some dynamic changes of few things in the notification itself. Instead of (relatively) simple BASH that almost everyone can use and hack something together, you'd end up atleast with Python.
> 
> I personally would say that BASH interface is much more user inviting - we could create few sample scripts so BFU can hack these for start. Much easier than Python/C/whatever else slightly_smiling_face

I thought a bit more about this and I think the bash interface with modifying env variables is fine. DBUS is probably a bit overkill for IPC, but I think pipes might be more suitable as a communication method that shared memory, since this is a one-directional communication scenario. Would you consider implementing pipes to replace the shared memory buffer?

Also, I see that you haven't made any tests yet. Could you do that as well?

**Comment by xoores** [comment]
> I thought a bit more about this and I think the bash interface with modifying env variables is fine. DBUS is probably a bit overkill for IPC, but I think pipes might be more suitable as a communication method that shared memory, since this is a one-directional communication scenario. Would you consider implementing pipes to replace the shared memory buffer?

Yeah, why not. I opted for the MMAP solely because the double-fork and this was the easiest solution at that moment :slightly_smiling_face:


> Also, I see that you haven't made any tests yet. Could you do that as well?

You mean those CI/CD checks? I cannot run them on my own as I'm the first time contributor. I did run your Valgrind test suite locally though and I'm using this version ever since and 90% of my notifications are modified by the script and everything seems to be OK :slightly_smiling_face:

File: src/notification.c [hunk 5]
```diff
@@ -47,130 +50,476 @@ struct _notification_private {
...
+                OVERRIDE_IF_DIFFERENT_STR(n, stack_tag, envValue);
+
+        } else if( STR_EQ("DUNST_ICON_PATH", envName) ) {
+                OVERRIDE_IF_DIFFERENT_STR(n, icon_path, envValue);
+
+        }  else if( STR_EQ("DUNST_URGENCY", envName) ) {
+                enum urgency u = URG_NONE;
+
+                // Dictionary would be nice, but hey...
+                if( STR_EQ("LOW", envValue) ) u = URG_LOW;
```
Comment (fwsmit) [inline src/notification.c:266]: There's a function for this: string_parse_enum (from option_parser.h). You can pass the urgency_enum_data (from settings_data.h) to it. The interface of the string_parse_enum function is a bit weird to handle errors (the return pointer isn't changed when an error occurs).

File: src/utils.h [hunk 6]
```diff
@@ -245,5 +245,23 @@ FILE * fopen_verbose(const char * const path);
...
+ *
+ * @param buffer Buffer to hexdump
+ * @param len Length of the buffer
+ */
+void hexdump(char * buffer, int len);
+
+//! Wrapper around hexdump() for string
+#define STR_HEXDUMP(STR)        hexdump(STR, strlen(STR))
+
+//! Escape LF and remove CR from string: \n -> \\n
```
Comment (fwsmit) [inline src/utils.h:260]: I think you shouldn't need to escape anything (it's C, not bash :p). Instead, try using NULL bytes for separation.

File: src/utils.h [hunk 7]
```diff
@@ -245,5 +245,23 @@ FILE * fopen_verbose(const char * const path);
...
  */
 void add_paths_from_env(GPtrArray *arr, char *env_name, char *subdir, char *alternative);
+
+/**
+ * Hexdumps buffer, 1 line = 8 bytes with character representation
+ *
+ * @param buffer Buffer to hexdump
+ * @param len Length of the buffer
+ */
+void hexdump(char * buffer, int len);
```
Comment (fwsmit) [inline src/utils.h:255]: Why don't we convert to int string instead? `printf("%i", val)`

File: src/rules.c [hunk 8]
```diff
@@ -59,6 +59,7 @@ void rule_apply(struct rule *r, struct notification *n)
                 n->markup = r->markup;
         if (r->icon_position != -1)
                 n->icon_position = r->icon_position;
+
```
Comment (fwsmit) [inline src/rules.c:62]: Try not to add random newlines.

File: .gitignore [hunk 9]
```diff
@@ -16,4 +16,4 @@ vgcore.*
 /dunstify
 /dunst.systemd.service
 /org.knopwob.dunst.service
-/test/test
+/test/test
```
Comment (fwsmit) [inline .gitignore:19]: Please try to undo this change to not clutter the diff

**Review by fwsmit — CHANGES_REQUESTED** [review]
I've submitted a few more comments. As a heads up, since there have been a few changes because of some failing tests and requested changes, I think it's best to squash the past commits into one, once you've implemented the feature. Then a few separate commits can be made to add tests and documentation. This squashing is important, because we need all commits to have working tests for bisecting. If you're not able to do that, I can do that for you as well.

File: src/utils.h (hunk 6 above)
Comment (xoores) [reply src/utils.h:260]: The escaping is done because some fields need the escaped form (format) and some interpret it (body, summary). If I didn't  escape it, BASH would escape the unescaped half and you'd probably mess up formatting of notifications :slightly_smiling_face:

File: src/utils.h (hunk 7 above)
Comment (xoores) [reply src/utils.h:255]: ?

Length is for binary buffer boundaries (so it can hexdump NULLs) - why would we need string representation of the buffer length?

File: src/notification.c (hunk 5 above)
Comment (xoores) [reply src/notification.c:266]: Ah, nice. I was trying to look for it but it ended up as quite a rabbithole when I looked on the configuration/parameters :laughing:

Will do :+1:

**Comment by fwsmit** [comment]
> > I thought a bit more about this and I think the bash interface with modifying env variables is fine. DBUS is probably a bit overkill for IPC, but I think pipes might be more suitable as a communication method that shared memory, since this is a one-directional communication scenario. Would you consider implementing pipes to replace the shared memory buffer?
> 
> Yeah, why not. I opted for the MMAP solely because the double-fork and this was the easiest solution at that moment slightly_smiling_face

I haven't looked at the implications of pipes and double forks, but I'm sure you can figure that out. If it's too complicated you can stick with mmap.


> Btw I also have some plans to implement Dunst module into Polybar so I can easily control Dunst without having to poll all the info through BASH script. For this I'd definitely go with DBUS slightly_smiling_face Maybe the interface already exist - did not really check yet.

There are some dbus interfaces already. Most of them are accesible with `dunstctl`. You can a look at the man page of dunstctl to see what's there. 

> 
> > Also, I see that you haven't made any tests yet. Could you do that as well?
> 
> You mean those CI/CD checks? I cannot run them on my own as I'm the first time contributor. I did run your Valgrind test suite locally though and I'm using this version ever since and 90% of my notifications are modified by the script and everything seems to be OK slightly_smiling_face Or do you mean enhancing current testsuite?

You are able to add tests. You can add them in `test/notification.c` or any other relevant test file. Test that rely on you being able to see if the notification is drawn correctly, you can add in `test/functional_test/test.sh`.

File: src/rules.c (hunk 8 above)
Comment (xoores) [reply src/rules.c:62]: yeah, that was a leftover :slightly_smiling_face: I removed it.

File: .gitignore (hunk 9 above)
Comment (xoores) [reply .gitignore:19]: > I haven't looked at the implications of pipes and double forks, but I'm sure you can figure that out. If it's too complicated you can stick with mmap.

I'll take a look. I think I will have to basically read the output twice. Not really a problem per se. :slightly_smiling_face: 


> There are some dbus interfaces already. Most of them are accesible with `dunstctl`. You can a look at the man page of dunstctl to see what's there.

I will :slightly_smiling_face: I will be looking for Dunst event that state has changed (paused/unpaused) at the minimum and possibly notification that amount of notifications on hold has changed.


> You are able to add tests. You can add them in `test/notification.c` or any other relevant test file. Test that rely on you being able to see if the notification is drawn correctly, you can add in `test/functional_test/test.sh`.

I'll take a look at it.

**Comment by xoores** [comment]
> I haven't looked at the implications of pipes and double forks, but I'm sure you can figure that out. If it's too complicated you can stick with mmap.

I'll take a look. I think I will have to basically read the output twice. Not really a problem per se. :slightly_smiling_face: 


> There are some dbus interfaces already. Most of them are accesible with `dunstctl`. You can a look at the man page of dunstctl to see what's there.

I will :slightly_smiling_face: I will be looking for Dunst event that state has changed (paused/unpaused) at the minimum and possibly notification that amount of notifications on hold has changed.


> You are able to add tests. You can add them in `test/notification.c` or any other relevant test file. Test that rely on you being able to see if the notification is drawn correctly, you can add in `test/functional_test/test.sh`.

I'll take a look at it.

**Comment by liskin** [comment]
> > I haven't looked at the implications of pipes and double forks, but I'm sure you can figure that out. If it's too complicated you can stick with mmap.
> 
> I'll take a look. I think I will have to basically read the output twice. Not really a problem per se. slightly_smiling_face

Only if you try to keep both the double fork and the `popen`. If you drop `popen` and `mmap` and replace it with some `pipe(2)` and `dup2(2)` calls, there'll be no need to read it twice.

**Comment by xoores** [comment]
> > > I haven't looked at the implications of pipes and double forks, but I'm sure you can figure that out. If it's too complicated you can stick with mmap.
> > 
> > 
> > I'll take a look. I think I will have to basically read the output twice. Not really a problem per se. slightly_smiling_face
> 
> Only if you try to keep both the double fork and the `popen`. If you drop `popen` and `mmap` and replace it with some `pipe(2)` and `dup2(2)` calls, there'll be no need to read it twice.

Yeah, I was also thinking about spawning a thread for this and share a memory directly. Dup2 would also be viable solution, perhaps even better as it will be completely isolated from Dunst. :+1:

**Comment by liskin** [comment]
> Yeah, I was also thinking about spawning a thread for this and share a memory directly. Dup2 would also be viable solution, perhaps even better as it will be completely isolated from Dunst. +1

A thread indeed sounds like unnecessary complexity. If we didn't need to set all those env vars, a simple

```
if want blocking:
   popen
   blocking read
   pclose
else:
   double fork and forget
```

would work just as well.

Unfortunately we need to set a bunch of env vars for the child process only, and popen doesn't provide a way to do this, so you'll probably end up reimplementing popen using pipe/dup2/fork/execve/waitpid. :-/

**Comment by fwsmit** [comment]
Hi @xoores,

Do you agree with the suggested changes and are you able to make the time for it?

**Comment by xoores** [comment]
> Unfortunately we need to set a bunch of env vars for the child process only, and popen doesn't provide a way to do this, so you'll probably end up reimplementing popen using pipe/dup2/fork/execve/waitpid. :-/

Yeah, but it's worth it to get the functionality :grin:


> Do you agree with the suggested changes and are you able to make the time for it?

Of course. I have some work related things that I need to do first so I think I should be able to do this during the weekend. :slightly_smiling_face:

**Comment by fwsmit** [comment]
@xoores Hi, would you still like to finish this PR? Otherwise I'm afraid I'm gonna have to close it.

**Comment by xoores** [comment]
@fwsmit Hi, yeah! I had quite a lot on my plate (health-wise) so I did not really had an opportunity to clean it up.

Of course I want to finish it, I will just have to refresh what I did & did not. 🙂

**Comment by fwsmit** [comment]
Unfortunate to hear about your health. I hope you're doing better now. Please only finish this if you feel like it. I just wanted to make sure it wasn't entirely forgotten.

As for the code, I think your approach should work, you just need to get the details right. In your first comment you mentioned something about backward compatibility. Please ensure your solution is 100% backward compatible. The dunst project isn't moving that fast anymore (see #1117 for an update) and I want to make sure it doesn't break on new releases.

**Comment by fwsmit** [comment]
For organizational purposes I'm going to close this pull request. If you or anyone else wants to work on this, feel free to reopen this pull request or start a new one (referencing this PR).
